Common++/header/PointerVector.h

```diff
@@ -58,7 +58,14 @@ namespace pcpp
 			for (const auto iter : other)
 			{
 				T* objCopy = new T(*iter);
-				m_Vector.push_back(objCopy);
+				try
```

Comment:
why do we need try-catch here? if we get an exception during the construction, the program should throw directly, Also, in `~PointerVector()`, the function already deletes the pointer.

Reply:
Because if the ctor throws you lose the pointer `objCopy` when the stack is unwinded and the memory allocated by `new T(*iter)` is leaked as you don't have any pointers referencing them after that.

Also did some tests and if an object ctor throws, the object's destructor is not called. All the initialized member fields do get their destructors called, but the object being constructed does not. So in actuality all the new deep copies of T are leaked if something fails during copy ctor.

Look at this code:
![image](https://github.com/seladb/PcapPlusPlus/assets/20269350/a72a9b77-ca5d-489e-8fbe-29d5116aeace)
![image](https://github.com/seladb/PcapPlusPlus/assets/20269350/0e578d6f-06e3-4ef1-8429-277acd5325db)

See how in the first version where it does not throw, `a` and `a2` are destructed, but in the second version where `A(const A& o)` throws, only `a` is destructed because `a2`'s copy ctor threw. This is essentially what will happen if PointerVector's copy ctor threw for any reason. The `m_Vector` is deconstructed properly, but you leak all the memory that the pointers being held within the vector point to.

Reply:
Also, yeah, mb, exception should be rethrown after delete.

Reply:
> Also, I am wondering if it ever has a chance to throw during `push()`?

`pushBack()` does not really have an issue with memory leaks internally, since it does no allocations directly. It is the caller's responsibility to have allocated the element T that is being pushed, and to safely do error handling if the push fails. and the `m_Vector.push_back()` will do its own recovery when throwing, so no need for interaction there.

Reply:
I understand your concern, but I am not sure if using the current two-level try-catch blocks is the best practice or not...
I agree that this patch ensures memory safety and `try` doesn't increase overhead. However, from the maintenance and conciseness perspective, I think we could find a better solution instead of using the two-level try-catch, what do you think?

Reply:
Technically a better way would be utilizing `unique_ptr` as a holding type for the internal vector. It would nicely handle most deallocations automatically. Tho the issue there ends up requiring making custom iterators if `PointerVector` is to keep its current API of returning pointers from the iterators.

Reply:
I agree it's not very easy to read, but I think it's fine because it's quite short